xblock/scorable.py

```diff
@@ -0,0 +1,158 @@
+"""
+Scorable.
+"""
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from collections import namedtuple
+import six
+
+
+Score = namedtuple('Score', ['earned', 'total'])
+
+
+class ScorableXBlockMixin(object):
+    """
+    Mixin to handle functionality related to scoring.
+
+    Provides the following public methods:
+
+        ScorableXBlockMixin.rescore(self, only_if_higher)
+            Calculate a new score and save it to the block.
+            `only_if_higher
+
+        ScorableXBlockMixin.allows_rescore(self) -> bool
+            Returns True if the block can be rescored.
+
+    Subclasses must define the following:
+
+        ScorableXBlockMixin.get_score(self)
+            Return the currently saved score.  Must not calculate a new score.
```

Comment:
These docstrings should specify weighted vs. raw. While theoretically an XBlock shouldn't know its own weight, the drag-and-drop v2 XBlock knows its own weight and stores its grade as a weighted value. While any block could choose to implement get_, set_, and calculate_score with either raw or weighted scores (as long as they were consistent), for our logging and debugging purposes I think that it is useful to know what data we are looking at and have it be consistent.